Typing errors composing commands created by `commandFactory`

**Bug**

When `tsconfig.json` is set to `module: "esnext"` and `target: "es6"` composing commands created by a `commandFactory` causes compilation errors.

Package Version: `latest`, `next`

**Code**

```ts
const store = new Store();
const createCommand = createCommandFactory<any>();
const composedCommand = createCommand(({ path }) => {
	return [
		add(path('foo'), 'bar')
	];
});
const command = createCommand((options) => {
	return [
		...composedCommand(options)
	]
});
const process = createProcess('test', [command]);
const executor = process(store);
executor({});
assert.strictEqual(store.get(store.path('foo')), 'bar');
```

**Expected behavior:**

Should compile without errors.

**Actual behavior:**

`tsconfig.json` with `module: "esnext"` and `target: "es6"` errors with:

```
error TS2488: Type must have a '[Symbol.iterator]()' method that returns an iterator.
```

--- a/src/stores/process.ts
+++ b/src/stores/process.ts
@@ -26,15 +26,36 @@ export interface CommandRequest<T = any, P extends object = DefaultPayload> exte
  * verify the type of multiple commands without explicitly specifying the generic for each command
  */
 export interface CommandFactory<T = any, P extends object = DefaultPayload> {
-	<R extends object = P>(command: Command<T, R>): Command<T, R>;
+	<R extends object = P>(command: AsyncCommandWithOps<T, R>): AsyncCommandWithOps<T, R>;
+	<R extends object = P>(command: SyncCommandWithOps<T, R>): SyncCommandWithOps<T, R>;
+	<R extends object = P>(command: AsyncCommand<T, R>): AsyncCommand<T, R>;
+	<R extends object = P>(command: SyncCommand<T, R>): SyncCommand<T, R>;
+}
+
+export interface AsyncCommandWithOps<T = any, P extends object = DefaultPayload> {
+	(request: CommandRequest<T, P>): Promise<PatchOperation<T>[]>;
+}
+
+export interface SyncCommandWithOps<T = any, P extends object = DefaultPayload> {
+	(request: CommandRequest<T, P>): PatchOperation<T>[];
+}
+
+export interface AsyncCommand<T = any, P extends object = DefaultPayload> {
+	(request: CommandRequest<T, P>): Promise<void>;
+}
+
+export interface SyncCommand<T = any, P extends object = DefaultPayload> {
+	(request: CommandRequest<T, P>): void;
 }
 
 /**
  * Command that returns patch operations based on the command request
  */
-export interface Command<T = any, P extends object = DefaultPayload> {
-	(request: CommandRequest<T, P>): Promise<PatchOperation<T>[]> | PatchOperation<T>[] | void | Promise<void>;
-}
+export type Command<T = any, P extends object = DefaultPayload> =
+	| SyncCommand<T, P>
+	| AsyncCommand<T, P>
+	| SyncCommandWithOps<T, P>
+	| AsyncCommandWithOps<T, P>;
 
 /**
  * Transformer function
@@ -132,7 +153,14 @@ export interface CreateProcess<T = any, P extends object = DefaultPayload> {
  * Creates a command factory with the specified type
  */
 export function createCommandFactory<T, P extends object = DefaultPayload>(): CommandFactory<T, P> {
-	return <R extends object = P>(command: Command<T, R>) => command;
+	function commandFactory<R extends object = P>(command: AsyncCommand<T, R>): AsyncCommand<T, R>;
+	function commandFactory<R extends object = P>(command: SyncCommand<T, R>): SyncCommand<T, R>;
+	function commandFactory<R extends object = P>(command: AsyncCommandWithOps<T, R>): AsyncCommandWithOps<T, R>;
+	function commandFactory<R extends object = P>(command: SyncCommandWithOps<T, R>): SyncCommandWithOps<T, R>;
+	function commandFactory<R extends object = P>(command: Command<T, R>): Command<T, R> {
+		return command;
+	}
+	return commandFactory;
 }
 
 /**

--- a/tests/stores/unit/process.ts
+++ b/tests/stores/unit/process.ts
@@ -795,6 +795,20 @@ const tests = (stateType: string, state?: () => MutableState<any>) => {
 			const processExecutor = process(store);
 			return processExecutor({});
 		});
+
+		it('should be able to compose commands created by a command factory', () => {
+			const createCommand = createCommandFactory<any>();
+			const composedCommand = createCommand(({ path }) => {
+				return [add(path('foo'), 'bar')];
+			});
+			const command = createCommand((options) => {
+				return [...composedCommand(options)];
+			});
+			const process = createProcess('test', [command]);
+			const executor = process(store);
+			executor({});
+			assert.strictEqual(store.get(store.path('foo')), 'bar');
+		});
 	});
 };
 

--- a/tests/stores/unit/process.ts
+++ b/tests/stores/unit/process.ts
@@ -180,20 +180,8 @@ const tests = (stateType: string, state?: () => MutableState<any>) => {
 			const createCommand = createCommandFactory<StateType>();
 
 			createProcess('test', [
-				createCommand(({ path }) => [
-					{
-						op: OperationType.ADD,
-						path: new Pointer(path('foo').path),
-						value: 3
-					}
-				]),
-				createCommand(({ path }) => [
-					{
-						op: OperationType.ADD,
-						path: new Pointer(path('a', 'b').path),
-						value: 'foo'
-					}
-				])
+				createCommand(({ path }) => [add(path('foo'), 3)]),
+				createCommand(({ path }) => [add(path('a', 'b'), 'foo')])
 			])(store)({});
 
 			assert.equal(store.get(store.path('a', 'b')), 'foo');

